nix/status-go/shared/build.nix

```diff
@@ -0,0 +1,141 @@
+{ stdenv, lib, fetchFromGitHub, buildGo114Package
+# Dependencies
+, xcodeWrapper
+, go, androidPkgs
+# metadata and status-go source
+, meta, source
+# build parameters
+, platform ? "android"
+, arch ? "386"
+, api ? "23" }:
+
+let
+  inherit (lib) attrNames getAttr strings concatStringsSep concatMapStrings;
+
+  removeReferences = [ go ];
+  removeExpr = refs: ''remove-references-to ${concatMapStrings (ref: " -t ${ref}") refs}'';
+
+  targetArchMap = rec {
+    "386" = "i686";
+    "arm" = "armv7a";
+    "arm64" = "aarch64";
+  };
+  ldArchMap = {
+    "386" = "x86";
+    "arm" = "arm";
+    "arm64" = "arm64";
+  };
+
+  # Shorthands for the built phase
+  targetArch = getAttr arch targetArchMap;
+  ldArch = getAttr arch ldArchMap;
+
+  osId = builtins.elemAt (builtins.split "\-" stdenv.hostPlatform.system) 2;
+  osArch = builtins.elemAt (builtins.split "\-" stdenv.hostPlatform.system) 0;
+
+  androidTarget = targetArch + "-linux-" + platform;
+
+  iosArchMap = {
+    "386" = "x86_64";
+    "arm" = "armv7";
+    "arm64" = "arm64";
+  };
+  iosSdk = if arch == "386" then "iphonesimulator" else "iphoneos";
+  iosArch = getAttr arch iosArchMap;
+  compilerFlags = if platform == "android" || platform == "androideabi" then
+  "\"-isysroot $ANDROID_NDK_HOME/toolchains/llvm/prebuilt/${osId}-${osArch}/sysroot -target ${androidTarget}${api}\""
+  else
+  "\"-isysroot $(xcrun --sdk ${iosSdk} --show-sdk-path) -miphonesimulator-version-min=7.0 -fembed-bitcode -arch ${iosArch}\"";
+
+  linkerFlags = if platform == "android" || platform == "androideabi" then
+  "\" --sysroot $ANDROID_NDK_HOME/toolchains/llvm/prebuilt/${osId}-${osArch}/sysroot \
+  -target ${androidTarget}${api} -v -Wl,-soname,libstatus.so\""
+  else
+  "\"--sysroot $(xcrun --sdk ${iosSdk} --show-sdk-path) -miphonesimulator-version-min=7.0 -fembed-bitcode -arch ${iosArch}\"";
+
+  buildMode = if platform == "ios" then "c-archive" else "c-shared";
+  libraryFileName = if platform == "ios" then "./libstatus.a" else "./libstatus.so";
+
+  compilerVars = if platform == "android" || platform == "androideabi" then
+      ''
+        export CC=$ANDROID_NDK_HOME/toolchains/llvm/prebuilt/${osId}-${osArch}/bin/clang
+        export CXX=$ANDROID_NDK_HOME/toolchains/llvm/prebuilt/${osId}-${osArch}/bin/clang++
+        export PATH=$ANDROID_NDK_HOME/toolchains/llvm/prebuilt/${osId}-${osArch}/bin:$PATH
+      '' else ''
+        export PATH=${xcodeWrapper}/bin:$PATH 
+        export CC=$(xcrun --sdk ${iosSdk} --find clang) 
+        export CXX=$(xcrun --sdk ${iosSdk} --find clang++)
+        '';
+  goOs = if platform == "android" || platform == "androideabi" then
+    "android"
+    else "darwin";
+
+  goArch = if platform == "android" || platform == "androideabi" then
+  if arch == "386" then "386" 
+  else if platform == "androideabi" then "arm "else "arm64"
+    else
+    if arch == "386" then "amd64" else 
+    if arch == "arm" then "arm" else "arm64";
```

Comment:
Christ...

Reply:
I was thinking more like:
```nix
{
  goArch =
    if isAndroid then (
      if arch == "386" then "386" else
      if platform == "androideabi" then "arm"
      else "arm64"
    ) else if isIOS then (
      if arch == "386" then "amd64" else 
      if arch == "arm" then "arm"
      else "arm64"
    ) else throw "Uknown Platform!";
}
```

Reply:
The `if platform == "androideabi" then "arm"` part makes me worried tho. Every other if is checking `arch`, but this one is checking `platform`. Some weird edge case?

Reply:
The `else "arm64"` is kinda ugly too. I'd prefer a clear case rather than `else`.